pkg/newstorage/newstorage.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package newstorage
+
+import "errors"
+
+// ErrStoreNotFound is returned when a store is not found.
+var ErrStoreNotFound = errors.New("store not found")
+
+// ErrDataNotFound is returned when data is not found.
+var ErrDataNotFound = errors.New("data not found")
+
+// StoreConfiguration represents the configuration of a store.
+type StoreConfiguration struct {
+	TagNames []string
+}
+
+// QueryOptions represents various options for Query calls in a store.
+type QueryOptions struct {
+	pageSize int
+}
+
+// QueryOption represents an option for a Query call in a store.
+type QueryOption func(opts *QueryOptions)
+
+// WithPageSize sets the maximum page size for data retrievals done within the Iterator returned by the Query call.
+// Paging is handled internally by the Iterator. Higher values may reduce CPU time and the number of database calls at
+// the expense of higher memory usage.
+func WithPageSize(size int) QueryOption {
+	return func(opts *QueryOptions) {
+		opts.pageSize = size
+	}
+}
+
+// Tag represents a Name + Value pair that can be associated with a k,v pair for querying later.
+type Tag struct {
+	Name  string
+	Value string
+}
+
+// Operation represents an operation to be performed in the Batch method.
+type Operation struct {
+	Key   string
+	Value []byte // A nil value will result in a delete operation
+	Tags  []Tag  // Optional
+}
+
+// Provider represents a storage provider.
+type Provider interface {
+	// OpenStore opens a store with the given name and returns a handle.
+	// If the store has never been opened before, then it is created.
+	OpenStore(name string) (Store, error)
+
+	// SetStoreConfig sets the configuration on a store.
+	// The store must be created prior to calling this method.
+	SetStoreConfig(name string, config StoreConfiguration) error
```

Comment:
It is debatable whether this should be generally config or specifically just the tags. I suppose we are doing this for future possibilities (but it would be nice to comment on if we have any envisioned future config properties beyond tags).

Reply:
I don't actually have any envisioned future config properties in mind. Do you think I should change it to just `TagNames`?